app/inpututils.cpp

```diff
@@ -2222,3 +2222,44 @@ double InputUtils::pixelDistanceBetween( const QPointF &p1, const QPointF &p2 )
 {
   return std::hypot( p1.x() - p2.x(), p1.y() - p2.y() );
 }
+
+bool InputUtils::layerHasGeometry( const QgsVectorLayer *layer )
+{
+  if ( !layer || !layer->isValid() )
+    return false;
+  return layer->wkbType() != Qgis::WkbType::NoGeometry && layer->wkbType() != Qgis::WkbType::Unknown;
+}
+
+bool InputUtils::layerVisible( QgsMapLayer *layer )
+{
+  QgsLayerTree *root = QgsProject::instance()->layerTreeRoot();
```

Comment:
Mind that `root` can also be null. Best to follow the previous implementation from: https://github.com/MerginMaps/mobile/blob/f7b663121365f62011f0d5fbeaa82c6ba1008b3f/app/activeproject.cpp#L388-L410